src/bridges/tokemak/DepositBridge.sol

```diff
@@ -0,0 +1,337 @@
+// SPDX-License-Identifier: GPL-2.0-only
+// Copyright 2020 Spilsbury Holdings Ltd
+pragma solidity >=0.8.4;
+
+import {SafeERC20} from '@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol';
+
+import {IDefiBridge} from '../../interfaces/IDefiBridge.sol';
+import {IERC20} from '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+import {IRollupProcessor} from '../../interfaces/IRollupProcessor.sol';
+
+import {AztecTypes} from '../../aztec/AztecTypes.sol';
+
+interface Ttoken is IERC20 {
+    function requestedWithdrawals(address account) external view returns (uint256, uint256);
+
+    function requestWithdrawal(uint256 amount) external;
+
+    function deposit(uint256 amount) external payable;
+
+    function withdraw(uint256 requestedAmount) external;
+
+    function withdraw(uint256 requestedAmount, bool asEth) external;
+
+    function underlyer() external view returns (address);
+}
+
+interface IManager {
+    function getCurrentCycleIndex() external view returns (uint256);
+
+    function getPools() external view returns (address[] memory);
+}
+
+contract DepositBridge is IDefiBridge {
+    using SafeERC20 for IERC20;
+
+    address public constant WETH = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
+
+    address public constant MANAGER = 0xA86e412109f77c45a3BC1c5870b880492Fb86A14;
+
+    address public immutable rollupProcessor;
+
+    uint256 internal constant MAX_UINT = type(uint256).max;
+
+    uint256 internal constant MIN_GAS_FOR_CHECK_AND_FINALISE = 83000;
+    uint256 internal constant MIN_GAS_FOR_FUNCTION_COMPLETION = 5000;
+
+    mapping(address => uint256) pendingWithdarawls;
+
+    mapping(address => address) tTokens;
+    mapping(address => address) assets;
+
+    error InvalidCaller();
+    error InvalidAssetType();
+    error InvalidAsset(address token);
+    error InvalidInput();
+
+    struct Interaction {
+        uint256 inputValue;
+        address tAsset;
+    }
+
+    uint256[] nonces;
```

Comment:
Using this nonces array is really not a good idea as @PhilWindle mentioned before. Looping over unlimited size arrays is an anti-pattern in solidity because it can brick the contracts (and it's expensive). But fixing this is quite straightforward. We can simple use linked list structure. The changes will have to be following:
1) Remove this nonces array,
2) add a new attribute to the Interaction struct referencing next (or previous) pending interaction,
3) save the nonce of the last processed interaction to the state so that you can start finalizing again in the next transaction,
3) loop over pending nonces using the new Interaction attribute.

You can check out my [path-registry](https://github.com/benesjan/path-registry/blob/a4811af1a9160860762657d59dfade1c0c5fd518/src/interfaces/IPathRegistry.sol#L32) project for inspiration on how to use linked list structure.